### DISCUSSION TITLE
Is there a way to tell spaCy that certain words are related to a certain number?

### DISCUSSION THREAD

**Discussion by vrunm** [explosion/spaCy#12876] [Help: Coding & Implementations]
I would like to ask if there's a way to tell spaCy that certain words in a sentence are related to certain number?
`
sentence = "The feed rate, aspirator rate, inlet and outlet temperature and air flow rate were approximately 3l/hr, 100%, 120C, 90C, and 357l/hr, respectively."`
From above, we know that feed rate is 3l/hr, aspirator rate is 100%, inlet temperature is 120C, outlet temperature is 90C, and finally, air flow rate is 357l/hr

What I'm trying to do is to extract information from scientific articles with regards to certain parameters that they use. I've learnt how to tag certain words and put a custom label on them. But I'm unsure how to link words in a sentence with numbers, especially with a sentence as complicated as above.

The code below shows what I've done to label certain entities in a sentence. I have a list of materials one file and a list of equipment in another file. Using these list, I extract information from a paragraph and then I can view it using pandas.

sentence = "The feed rate, aspirator rate, inlet and outlet temperature and air flow rate were approximately 3l/hr, 100%, 120C, 90C, and 357l/hr, respectively."

From above, we know that feed rate is 3l/hr, aspirator rate is 100%, inlet temperature is 120C, outlet temperature is 90C, and finally, air flow rate is 357l/hr

The code below shows what I've done to label certain entities in a sentence. I have a list of materials one file and a list of equipment in another file. Using these list, I extract information from a paragraph and then I can view it using pandas.

```

import json
import spacy
from spacy import displacy
from spacy.matcher import Matcher

nlp = spacy.load("en_core_web_sm")
matcher = Matcher(nlp.vocab)



def load_data(file):
    with open(file,"r", encoding = "utf-8") as f:
        data = json.load(f)
    return (data)


def create_training_data(file1,label1,file2,label2):
    data1 = load_data(file1)
    data2 = load_data(file2)
    patterns = []
    for item in data1:
        pattern = {
            "label":label1,
            "pattern":item
            }
        patterns.append(pattern)
    for item in data2:
         pattern = {
             "label":label2,
             "pattern":item
                 }
         patterns.append(pattern)
    return (patterns)
```
I would like to do the same for parameters. Where I can extract the parameter name, and value that comes with it. Is it possible?

**Comment by Alyarin** [comment]
SpaCy comes with a dependency parser that you can use for this kind of thing.

**Reply by Alyarin** [reply]
Look up dependency trees, and try to figure out how exactly they work.

**Comment by vrunm** [comment]
@Alyarin After looking at dependency parser, I realised that I can use the patterns entity ruler
```
import spacy

#Build upon the spaCy Small Model
nlp = spacy.blank("en")
ruler = nlp.add_pipe("entity_ruler")
patterns = [
    {"label": "PARAMETER", "pattern": [{"TEXT":{"REGEX":r"(inlet temperature|outlet temperature|inlet|outlet)"}}]},    
    {"label": "TEMP", "pattern": [{"TEXT":{"REGEX":"\d+(C|K|F)"}}]},
    {"label": "aspirator rate", "pattern": [{"TEXT":{"REGEX":r"\d{1,5}"}},{"LOWER":r"%"}]},
    {"label": "feed rate", "pattern": [{"TEXT":{"REGEX":r"\d{1,5}[ml/min]"}},{"LOWER":r"ml/min"}]}
    
]

ruler.add_patterns(patterns)

text = "The feed rate, aspirator rate, inlet and outlet temperature and air flow rate were approximately 3 ml/min, 100%, 120C and 90C and 357 l/hr, respectively."

##Process the text with spaCy
doc = nlp(text)

# Iterate over the entities in the document
for ent in doc.ents:
    print (ent.label_,ent.text)
```
This gives a very awkward output of
```
PARAMETER inlet
PARAMETER outlet
aspirator rate 100%
TEMP 120C
TEMP 90C

```
 Seems that I was unable to get the feed rate or the parameter "outlet temperature"

I'm only trying things out to see if I can separate them like this. May I ask for your opinions on how to use REGEX to extract words like "ml/min","L/hr" or anything with special characters.
Sometimes scientific articles have different ways of writing. Some will write "inlet air temperature" while some will write "inlet temperature".  Can you use the  REGEX to encompass all these varieties? So I can assign PARAMETER --> outlet air temperature
